lib/pacemaker/pcmk_sched_colocation.c

```diff
@@ -1474,19 +1474,50 @@ pcmk__apply_coloc_to_priority(pcmk_resource_t *dependent,
  * \internal
  * \brief Find score of highest-scored node that matches colocation attribute
  *
- * \param[in] rsc    Resource whose allowed nodes should be searched
- * \param[in] attr   Colocation attribute name (must not be NULL)
- * \param[in] value  Colocation attribute value to require
+ * \param[in]     colocation  Colocation constraint being applied
+ * \param[in,out] rsc         Resource whose allowed nodes should be searched
+ * \param[in]     attr        Colocation attribute name (must not be NULL)
+ * \param[in]     value       Colocation attribute value to require
  */
 static int
-best_node_score_matching_attr(const pcmk_resource_t *rsc, const char *attr,
+best_node_score_matching_attr(const pcmk__colocation_t *colocation,
+                              pcmk_resource_t *rsc, const char *attr,
                               const char *value)
 {
+    GHashTable *allowed_nodes_orig = NULL;
     GHashTableIter iter;
     pcmk_node_t *node = NULL;
     int best_score = -PCMK_SCORE_INFINITY;
     const char *best_node = NULL;
 
+    if ((colocation != NULL) && (rsc == colocation->dependent)
+        && pcmk_is_set(colocation->flags, pcmk__coloc_explicit)
+        && pcmk__is_group(rsc->private->parent)
+        && (rsc != rsc->private->parent->private->children->data)) {
+        /* The resource is a user-configured colocation's explicit dependent,
+         * and a group member other than the first, which means the group's
+         * location constraint scores were not applied to it (see
+         * pcmk__group_apply_location()). Explicitly consider those scores now.
+         *
+         * This does leave one suboptimal case: if multiple members of a group
```

Comment:
It would be nice to have a regression test for this, with a note explaining "this is acceptable but suboptimal." And maybe a TODO flag here, mostly to help us find it if this ever becomes a problem.

It seems like this would also apply if both the group itself and any single member were explicitly colocated with another resource.

Reply:
makes sense, though I'll leave a regression test as an exercise for the reader ;)

Reply:
As a math student, that phrase is traumatic to me